### PR TITLE
Add web-platform-tests/web-animations test coverage to the polyfill

### PR DESCRIPTION
**Pull request by alancutter**
This adds automation for running [web-platform-tests/web-animations](https://github.com/w3c/web-platform-tests/tree/master/web-animations) tests to `grunt test` and `grunt debug`.
Existing failures are captured in `test/web-platform-tests-expectations.js`.

This change builds upon: https://github.com/web-animations/web-animations-next/pull/450

### CODE REVIEW

File: target-config.js [hunk 1]
```diff
@@ -119,7 +119,8 @@
       webAnimations1BonusSrc: webAnimations1BonusSrc,
       webAnimationsNextSrc: [],
       src: scopeSrc.concat(sharedSrc).concat(webAnimations1Src).concat(webAnimations1BonusSrc),
-      test: webAnimations1Test,
+      polyfillTests: webAnimations1PolyfillTests,
+      runWebPlatformTests: false,
```
Comment (suzyh) [inline target-config.js]: Why false?

File: Gruntfile.js [hunk 2]
```diff
@@ -186,17 +186,45 @@ module.exports = function(grunt) {
...
+      karmaConfig.files = ['test/karma-mocha-setup.js'].concat(config.src, config.polyfillTests);
+    }).then(function(success) {
+      if (!config.runWebPlatformTests) {
+        done(success);
+        return;
+      }
+      mochaSuccess = success;
+      return runKarma(function(karmaConfig) {
+        configCallback(karmaConfig);
+        karmaConfig.client.testList = grunt.file.expand('test/web-platform-tests/web-animations/**/*.html');
```
Comment (suzyh) [inline Gruntfile.js]: Does the ** do recursive expansion?

File: test/resources/testharnessreport.js [hunk 3]
```diff
@@ -406,17 +406,7 @@ function dump_test_results(tests, status) {
...
-    var scripts = document.getElementsByTagName('script');
-    var thisPath = scripts[scripts.length - 1].src;
-    var importPath = thisPath.replace(/([^\/]+\/){2}testharnessreport.js$/, 'web-animations-next.dev.js');
-    if (thisPath == importPath) {
-        throw new Error("Failed to generate polyfill library path.");
-    }
-    document.write('<script src="' + importPath + '"></script>');
-})();
-
+document.write('<script src="/polyfill/web-animations-next.dev.js"></script>');
```
Comment (suzyh) [inline test/resources/testharnessreport.js]: <3
So much nicer.

File: target-config.js (hunk 1 above)
Comment (alancutter) [inline target-config.js]: Hmm, this should probably be a list of files that we think native Chrome should pass instead, will take a look.
It's lame that webAnimations1 actually means a subset of level 1, if the name were accurate then we'd by trying to make it pass all the tests.

File: Gruntfile.js (hunk 2 above)
Comment (alancutter) [inline Gruntfile.js]: Yes, it works the same as Bash.

File: test/blink/2-keyframes-with-offsets.html [hunk 4]
```diff
@@ -1,6 +1,6 @@
 <!DOCTYPE html>
-<script src="testharness/testharness.js"></script>
-<script src="testharness/testharnessreport.js"></script>
+<script src="../resources/testharness.js"></script>
```
Comment (suzyh) [inline test/blink/2-keyframes-with-offsets.html]: Can you set this up as /resources or something now you have the 'proxies' config stuff?

File: target-config.js (hunk 1 above)
Comment (alancutter) [inline target-config.js]: On second thought I'd prefer to leave this to a separate patch to deal with the issue of expectation inheritance.

File: test/blink/2-keyframes-with-offsets.html [hunk 5]
```diff
@@ -1,6 +1,6 @@
 <!DOCTYPE html>
-<script src="testharness/testharness.js"></script>
-<script src="testharness/testharnessreport.js"></script>
+<script src="../web-platform-tests/resources/testharness.js"></script>
```
Comment (suzyh) [inline test/blink/2-keyframes-with-offsets.html]: I think these three script imports can now be
src="/resources/testharness.js"
src="/resources/testharnessreport.js"
and, with one more line in your karma config,
src="/resources/keyframes-test.js"
Is that right?

Comment (alancutter) [inline test/blink/2-keyframes-with-offsets.html]: Yes except for that last one. We can't have /resources/ pointing to two different places, it can just remain a relative path without any issues.

Comment (alancutter) [inline test/blink/2-keyframes-with-offsets.html]: Done.

Comment (suzyh) [inline test/blink/2-keyframes-with-offsets.html]: My concern is that it may be confusing to come into the code and see some scripts imported from "/resources" and some from "resources". I think you could add the third one as well if you specified '/resources/keyframes-test.js' as an entry in the karma config, but that may very well be undesirable. An alternative may be to rename one of them so the imports don't just differ in the presence of a single slash.

File: test/karma-testharness-adapter.js [hunk 6]
```diff
@@ -0,0 +1,181 @@
...
+      description: '',
+      success: message == '',
+      log: message ? [message] : [],
+    };
+  }
+
+  function checkConfig(config) {
+    var requiredProperties = {
+      testURLList: '<Array of test URLs>',
+      skip: '<Object maping skipped test URLs with the reason for skipping>',
```
Comment (suzyh) [inline test/karma-testharness-adapter.js]: Typo: maping -> mapping
Nit: with -> to

File: test/karma-testharness-adapter.js [hunk 7]
```diff
@@ -0,0 +1,181 @@
...
+      expectedFailures: '<Object mapping test URLs to expected inner test failures>',
+      flakyFailureIndicator: '<String used in expectedFailures to indicate non deterministic failure messages>',
```
Comment (suzyh) [inline test/karma-testharness-adapter.js]: I expect "flaky" to mean "sometimes passes, sometimes fails" but the description suggests you mean "fails but the failure message can vary from run to run". Can you clarify which you mean? Or is it both? Consider changing name and/or description to clarify.

File: test/karma-testharness-adapter.js [hunk 8]
```diff
@@ -0,0 +1,181 @@
...
+    };
+    var errorMessage = '';
+    if (!config) {
+      errorMessage = 'Missing testharnessTests config options on Karma\'s config.client.\n';
+    } else {
+      for (var property in requiredProperties) {
+        if (!(property in config)) {
+          errorMessage += 'Missing property ' + property + '\n';
```
Comment (suzyh) [inline test/karma-testharness-adapter.js]: Please extend the error message slightly to explain where the property is missing from.

File: test/blink/2-keyframes-with-offsets.html (hunk 5 above)
Comment (alancutter) [inline test/blink/2-keyframes-with-offsets.html]: Note that having local resource folders is used in web-platform-tests as well, not just Blink imported tests. I don't see why using relative paths should be confusing, adding extra indirection via config files is much harder to read to me.

File: test/karma-testharness-adapter.js (hunk 7 above)
Comment (alancutter) [inline test/karma-testharness-adapter.js]: I mean "fails but the failure message can vary from run to run". I suppose it'd be better if it meant can pass/fail, do you think so?

File: test/karma-testharness-adapter.js (hunk 8 above)
Comment (alancutter) [inline test/karma-testharness-adapter.js]: Done.

File: test/karma-testharness-adapter.js (hunk 6 above)
Comment (alancutter) [inline test/karma-testharness-adapter.js]: Done.

File: test/karma-testharness-adapter.js [hunk 9]
```diff
@@ -0,0 +1,186 @@
...
+    karma.info({total: config.testURLList.length});
+
+    var testName = document.createElement('div');
+    document.body.appendChild(testName);
+    var iframe = document.createElement('iframe');
+    iframe.style.width = 'calc(100vw - 30px)';
+    iframe.style.height = 'calc(100vh - 60px)';
+    document.body.appendChild(iframe);
+
+    function runRemainingTests() {
```
Comment (suzyh) [inline test/karma-testharness-adapter.js]: I don't understand the control flow for this function. Why is it called recursively? And can the function be defined outside of the definition of karma.start so that it's easier to read / easier to see what's inside runRemainingTests and what's not?

File: test/blink/2-keyframes-with-offsets.html (hunk 5 above)
Comment (suzyh) [inline test/blink/2-keyframes-with-offsets.html]: Fair. No worries then, this is fine.

File: test/karma-testharness-adapter.js (hunk 7 above)
Comment (suzyh) [inline test/karma-testharness-adapter.js]: I don't know if we have tests that are flaky in the sense of being able to pass or fail. Perhaps allow for both:
nonDeterministicIndicator: '<String used in expectedFailures to indicate non-deterministic test outcome (pass/fail, or fail with different error messages).>'

File: CONTRIBUTING.md [hunk 10]
```diff
@@ -1,11 +1,12 @@
 ## Developer setup instructions
 
 1. `git clone git@github.com:web-animations/web-animations-next.git`
+1. `git submodule update --init --recursive` (Necessary for running tests.)
 1. Install [node](https://nodejs.org/en/) and make sure `npm` is in your $PATH
 1. Run `npm install` in the respository to pull in development dependencies.
 1. Run `npm install -g grunt grunt-cli` to get the build tools for the command line.
 1. Run `grunt` to build the polyfill.
-1. Run `grunt test` to run polyfill tests using the build.
+1. Run `grunt test` to run polyfill and web-platform-tests tests.
```
Comment (suzyh) [inline CONTRIBUTING.md]: Please add a paragraph or two explaining what the setup is and how it works, so it's easier for a newbie to see how to modify it later. My understanding is roughly: grunt starts up karma, which starts up a browser, which executes some mocha tests and some other tests that consist of testharness wrapped inside a mocha adapter.

Comment (alancutter) [inline CONTRIBUTING.md]: Added a description of what happens, I'm a bit worried that it's too specific and will get outdated accidentally.

File: test/karma-testharness-adapter.js (hunk 9 above)
Comment (alancutter) [inline test/karma-testharness-adapter.js]: It was being defined inside karma.start so it could closure in a bunch of variables. Moved it outside for comparison.
It's called recursively because async for loops in Javascript aren't available cross browser yet, each item needs to wait for testharness.js' completion callback before going onto the next one.

File: test/karma-testharness-adapter.js (hunk 7 above)
Comment (alancutter) [inline test/karma-testharness-adapter.js]: We currently don't need such a thing, adding it would make the expected failure logic even more complex for no functional gain. I'd much prefer to have just one notion of flakiness for simplicity, I suspect pass/fail is the clearer definition.

Comment (alancutter) [inline test/karma-testharness-adapter.js]: Made flaky mean pass/fail.

File: CONTRIBUTING.md [hunk 11]
```diff
@@ -64,4 +65,24 @@
...
-    git push web-animations-js HEAD:refs/heads/master
\ No newline at end of file
+    git push web-animations-js HEAD:refs/heads/master
+
+## Testing architecture
+
+This is an overview of what happens when `grunt test` is run.
+
+1. Polyfill tests written in mocha and chai are run.
+1.1 grunt creates a karma config with mocha and chai adapters.
```
Comment (suzyh) [inline CONTRIBUTING.md]: This markup doesn't seem to be correct; the preview is not creating the sublists correctly.

File: CONTRIBUTING.md (hunk 10 above)
Comment (suzyh) [inline CONTRIBUTING.md]: I think it helps, and the fact that it's in version control with the configuration itself means that we can at least see how out of date it is at any point in time. Let's go with it, and edit it later if it becomes too cumbersome.

File: CONTRIBUTING.md (hunk 11 above)
Comment (alancutter) [inline CONTRIBUTING.md]: Fixed.

**Comment by suzyh** [comment]
LGTM. Thanks for all your work on this!

**Comment by alancutter** [comment]
Thanks for the thorough review.
